### DISCUSSION TITLE
Calendar

### DISCUSSION THREAD

**Discussion by ME-MarvinE** [CommunityToolkit/Maui#265] [New Feature Discussions]
I think a Calendar control would be a great addition to the toolkit. Examples of its use would be making your own date picker, making your own event calendar, or just a nice visual way of displaying data that has dates.

There's a ~~working solution for Xamarin Forms~~ [working solution for Xamarin Forms and .NET MAUI](https://github.com/ME-MarvinE/XCalendar) on my GitHub along with a sample and an in-depth [Wiki](https://github.com/ME-MarvinE/XCalendar/wiki).

**Comment by jfversluis** [comment]
Great idea Marvin! 😝 

Just for completeness sake I will link the Xamarin Community Toolkit Calendar PR here: https://github.com/xamarin/XamarinCommunityToolkit/pull/1331

We've closed that, but there is still probably some good input in there so we can take the best of all calendar components out there and create the best one!

What I personally was wondering from quickly using it... Why did you choose to have a `SelectedDate` and `SelectedDates`. In my opinion that will always cause confusion for users because one of them isn't getting updated depending on what the `SelectionMode` is. Is there any reason we should not always have `SelectedDates` but it will just have 1 value when `SelectionMode=Single`?

Do you have any more ideas about APIs that you might want to change or add or maybe even remove?

**Comment by ME-MarvinE** [comment]
> What I personally was wondering from quickly using it... Why did you choose to have a SelectedDate and SelectedDates. In my opinion that will always cause confusion for users because one of them isn't getting updated depending on what the SelectionMode is. Is there any reason we should not always have SelectedDates but it will just have 1 value when SelectionMode=Single?

~~This is what I wanted to do originally. However I tried to use the default Xamarin Forms controls as a guide for implementation in order to make the Calendar more intuitive. I saw that `CollectionView` had a `SelectedItem` and `SelectedItems` property so I thought it would be best if I just copied it with `SelectedDate` and `SelectedDates`.~~ **Implemented in 2.0.1**

> Do you have any more ideas about APIs that you might want to change or add or maybe even remove?
<details>
<summary>Answer</summary>

* **Add** flow directions for dates. For example dates could go from: top-left to top-right, top-right to top-left,  bottom-right to bottom-left or bottom-left to bottom-right.
* ~~**Add** support for culture~~ **Already achievable using `CalendarView`'s templates**
* ~~**Change** implementation of `CalendarView` from `CollectionView` to `Grid`. This may provide better performance or make it easier to customise days due issues such as:~~ **View Grid-based CalendarView branch [here](https://github.com/ME-MarvinE/XCalendar/tree/grid-based-daysview). Negligible performance difference, although the animated swipable calendar example is much faster.**
    * dotnet/maui#3885
    * Xamarin/Xamarin.Forms#11053
    * https://github.com/xamarin/Xamarin.Forms/issues/7115#issuecomment-519836059 (An extra view in a control like this causes major performance impacts)

* ~~**Change** Have a default implementation for a CalendarDayView or have customisation in its own class or control so that when making a template, you don't have to bind everything you want again. There are 5 states and each has a TextColor property. You could see how it would quickly get out of hand if there was also a BackgroundColor or BorderColor property for each state.~~ **Implemented in 2.1**

</details>

**Comment by ME-MarvinE** [comment]
There is now a [pre-release](https://www.nuget.org/packages/XCalendar.Maui) of [XCalendar](https://github.com/ME-MarvinE/XCalendar) for .NET MAUI.

**Comment by ME-MarvinE** [comment]
XCalendar.Maui is now out of pre-release.

Although, it's very slow compared to Xamarin Forms and I don't know why. XCalendar could probably be faster with MAUI Handlers and Custom Renderers but I know nothing about platform-specific controls.

XCalendar's logic is also now platform independent via the 'Calendar' class in 'XCalendar.Core'. What would happen to XCalendar.Forms and XCalendar.Core if XCalendar was added to the Maui Community Toolkit? Should they be added somewhere else?

**Comment by yairp03** [comment]
Any update on this one?

**Comment by EugenM88** [comment]
Greetings!

.net MAUI still has no calendar control as of today, and this great communitytoolkit unfortunately also lacks such much needed feature.
I second the petition for feature to be implemented as im sure im not the only seeing the potential in such feature.
WinUI2 and WinUI3, WPF and WinForms already have a control for such a necessary thing, why not .net Maui?

Best Regards!

**Comment by HobDev** [comment]
XCalendar is a nice addition to Maui Community toolkit

**Comment by TheCodeTraveler** [comment]
Thanks @ME-MarvinE! Congrats on the release of XCalendar.Maui! https://www.nuget.org/packages/XCalendar.Maui

We discussed this in the [November Standup](https://www.youtube.com/watch?v=a6LBam_9q-0), and we agree that your existing implementation is great and we encourage you to continue the great work! 

There are also Calendar controls created by [SyncFusion](https://help.syncfusion.com/maui/calendar/getting-started) and [Telerik](https://www.telerik.com/maui-ui/calendar) that work great too!

**Reply by MichaelDenwood** [reply]
Will @ME-MarvinE be contributing the calendar to the Maui Community Toolkit?  We would like to see an officially supported solution as too many times developers create great projects but stop supporting them as they don't have time.  Other times the repo and all traces of the product vanish.  

I would encourage this project team to consider a permeant solution because not every company allows the purchase of third-party components. 

It is also noteworthy to mention the Xamarin Community Toolkit also offered a Calendar.   Maybe it could it be ported to Maui?

**Reply by bijington** [reply]
I can see that the current package is being actively maintained as there was a release 6 days ago. While it would be great for the toolkit to encompass more features, it puts an extra burden on the team. Of course we are a community toolkit so if you would be willing to contribute towards features like this then please do let us know.

I'm not sure the Xamarin Community Toolkit did offer a calendar control. Where did you find that?

**Reply by MichaelDenwood** [reply]
Oh you are correct, it was part of Xamarin.  I mistakenly thought it was in the Xamarin Community Toolkit.  
https://learn.microsoft.com/en-us/xamarin/android/user-interface/controls/calendar

My point about things not being maintained is times we have been burnt using private contributors' libraries in the past.  I was not saying it would be the case in this situation, but it is possible, so our company frowns upon it.  I would be happy to contribute but simply don't have the time at the moment but would love to in the future.  I will ask Microsoft about this since Xamarin officered this control out of the box and we are required to migrate to Maui as our Xamarin app will stop working in the next few months if it is not ported to Maui.

**Reply by bijington** [reply]
I understand companies don't like to rely on single developers for the risk of them disappearing. The other issue is that they like to depend on them or others to obtain some value for no cost. I'm not saying this is the case for your company but it is a typical reason why the single developers get burnt out.

There are some potential approaches that your company could introduce to mitigate the risk though:
- sponsor developers if they offer it, this helps them to afford to do the work in their professional time and not just free time
- dedicate some time for their developers to contribute to the library they are consuming
- purchase a license for a commercial offering from the vendors above
- fork the repo, so if the repo disappears you have a backup

That list is in order of which I would advocate most for.

Also a final note, I doubt your app will stop working in May 2024 (unless there is some context I don't know), Microsoft will just stop supporting Xamarin from May 2024. It is still very good to be migrating now though

**Reply by jfversluis** [reply]
Another quick note on this, just to be sure...

> We would like to see an officially supported solution as too many times developers create great projects but stop supporting them as they don't have time. 

The Community Toolkit is a _community_ project just as well. It might have a bit more organization and it's closer to the .NET MAUI sources for sure, but there is nothing more official about it than any other package basically. Just to set the right expectations
